### DISCUSSION TITLE
How do I get rid of spikes in my simulation result?

### DISCUSSION THREAD

**Discussion by steam-simulator** [idaholab/moose#27077] [Q&A Modules: Porous Flow] [answered]
Hello,
My simulation was successful with the expected result. However, the result show too many spikes. How do I get rid of these spikes? I’m using `IterationAdaptiveDT` with `dt_min = 1e-3`. When I removed the IterationAdaptiveDT, I don't see the spikes, but my results deviated slightly from what I would expect. I want to use the iterative adaptive time-stepping. Any suggestion on how I can do this without the spikes? Thanks.

```
[Executioner]
  type = Transient
  solve_type = Newton
  end_time = 3e3
  nl_abs_tol = 1e-9
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 1e-3          
    optimal_iterations = 10
  []
[]
```

**Comment by GiudGiud** [comment]
Hello

What kind of spikes are they? In time? or in the spatial solution?

My worry with spikes would be that you have false convergence happening. Change `nl_abs_tol = 1e-9` this back to the default. Removing the line will do that

Guillaume

**Comment by steam-simulator** [comment]
Hello,

The spikes occur in time. After changing the tolerance to the default value, I got a convergence error:
```
*** ERROR ***
The following error occurred in the object "IterationAdaptiveDT", of type "IterationAdaptiveDT".

Solve failed and timestep already at dtmin, cannot continue!
```

**Reply by GiudGiud** [reply]
ok can you turn on automatic scaling then set a reasonably tight tolerance. 
automatic scaling will normalize the tolerances

**Comment by steam-simulator** [comment]
Hello @GiudGiud, my simulation fails to converge after setting a tight tolerance and using automatic scaling.  Any other suggestion to make the IterationAdaptiveDT work without the spike? Thanks.

**Reply by GiudGiud** [reply]
I dont think the time stepper is the problem here. They simply create the time grid. They both can be made to have a sufficiently fine time step for your problem.
Two things could be happening here:
- the time steps are not well selected and in the first oscillating case you were violating a CFL-like condition, causing stability issues
- the convergence criteria were loosened and the results started oscillating because they were not converged.

Could you please paste the executioner block?

**Reply by steam-simulator** [reply]
Hi @GiudGiud, here is the executioner block that I used: 
```
[Executioner]
  type = Transient
  solve_type = Newton
  end_time = 3e3
  nl_abs_tol = 1e-11
  [TimeStepper]
  type = ConstantDT
  dt = 1
  []
[]
```

**Reply by GiudGiud** [reply]
the executioner looks fine to me. 
nl_abs_tol of 1e-11 is probably fine enough. I like to use `automatic_scaling = true` in that block to normalize the convergence criteria but I see you already have manual scaling in the simulation so it's probably already fine.

I think looking at the IterativeAdaptiveDT time steps might show you if they were too big for the problem

**Reply by steam-simulator** [reply]
Hello @GiudGiud, how do I output the iterativeAdaptiveDT time steps? Should I use the output system to do this?

**Reply by cpgr** [reply]
You can use a TimeStepSize post processor to save each time step into a csv file

**Reply by steam-simulator** [reply]
Hello @GiudGiud @cpgr, thanks for the feedback. Now I know why I'm getting spikes in the result. Is there a way to maintain the time steps in a simulation when using `constantDT` without the time step fluctuating at certain times? For example, I set dt=4 using `constantDT`, but the time step is not constant during the simulation. It fluctuates from 4 to 2 at certain times, corresponding to the spikes. Is there a way (or time stepper) to force the time step to remain at a specific value throughout the simulation even if any fluctuation occurs?

**Reply by GiudGiud** [reply, marked as answer]
Hello

The fluctuations are probably due to time steps failing? What does the log say?
I think you are having convergence issues here. You need to work on getting clean nonlinear solve convergence. Maybe that will force a reduction of the time step. Maybe you can work with the executioner parameters (maybe automatic scaling?) to get better convergence

Guillaume

**Reply by steam-simulator** [reply]
Thanks for the tips, @GiudGiud
